File: 98. Validate Binary Search Tree/ValidateBinarySearchTree.py
#@author:leacoder
#@des:  中序遍历  验证二叉搜索树
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        inorderlist = self.inorder(root) #中序遍历后 如果是二叉搜索树 那么 结果必然是递增有序的 
        return inorderlist == list(sorted(set(inorderlist))) #set 集合用于去重  有重复的数那么必然不是二叉搜索树
        
    def inorder(self,root): #中序遍历 形成 左 根 右形式
        if root is None:
            return []
        return self.inorder(root.left) + [root.val] + self.inorder(root.right)
    
    
    '''
    sort() 对list本身进行排序,改变list的值。sort()只能对list排序。
    sorted() 产生一个新的list，不改变list的值。sorted()可以对iterable对象排序
    集合（set）是一个无序的不重复元素序列。
    可以使用大括号 { } 或者 set() 函数创建集合，注意：创建一个空集合必须用 set() 而不是 { }，因为 { } 是用来创建一个空字典。

File: 98. Validate Binary Search Tree/ValidateBinarySearchTree.java
/*
 *@author:leacoder
 *@des:  递归比较min max,  验证二叉搜索树
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        Integer min = null;
        Integer max = null;
        
        return isValid(root,min,max);
    }
    public boolean isValid(TreeNode root,Integer min,Integer max){ //min 下界   max上界
        if(root == null) return true;
        if(min!=null && root.val <=min) return false; 
        if(max!=null && root.val >=max) return false;  //当前root节点值 必须在 min 和 max之间
        
        //注意下边函数式子     分别递归检测   root的左子树（min下界不关心，其上界必须是root的值） 与 root右子树（max上界不关心，其下界界必须是root的值）
        return isValid(root.left,min,root.val)&&isValid(root.right,root.val,max);
    }
}